content/blog/2023/2023-09-04-nodejs-rate-limiter.md

```diff
@@ -0,0 +1,250 @@
+
+Middleware functions are commonly used to perform various tasks related to request processing, such as authentication, logging, data validation, and rate limiting.
+
+The rate limiter configuration will be provided as middleware from the ratelimit.js file.
+
+### Step 2: Install the Application Dependencies
+To install the necessary packages for our application, Run the following command:
+
+```bash
+npm install express express-rate-limiter
+```
+
+Where:
+* **express:** is a web application framework for Node.js. It simplifies the process of building robust, scalable, and performant web applications and APIs.
+* **express-rate-limiter:** is a middleware for rate limiting in Express.js applications. It allows us to control the rate at which requests are allowed to our Express routes.
+
+### Step 3: Starting Our Node.js Server
+Now, we'll create basic APIs for our application and start the Node.js server.
+
+For this, copy and paste the following code into the `app.js` file:
+```js
+const express = require("express");
+const rateLimitMiddleware = require("./middlewares/ratelimit");
+const app = express();
+
+// A simple API route
+app.get("/api/blog", (req, res) => {
+  res.send({
+    success: true,
+    message: "Welcome to our Blog API Rate Limiter Project 🎉",
+  });
+});
+
+app.get("/api/blog/post", (req, res) => {
+  res.send({
+    success: true,
+    author: "Mike Abdul",
+    "title": "Creating NodeJs Rate Limiter",
+    "post": "..."
+  });
+});
+
+const PORT = process.env.PORT || 5000;
+app.listen(PORT, () => {
+  console.log(`Server running on port ${PORT}`);
+});
+```
+With this code, we have successfully set up two API routes `/api/blog` and `/api/blog/post` both of which are not rate-limited. 
+We can start the application server by executing the command
+```bash
+node app.js
+```
+When testing the APIs with tools like Postman or a web browser, we will notice that there are currently no limits on the number of calls we can make to these endpoints.
+
+{{% image alt="non-rate-limited-api" src="images/posts/nodejs-rate-limiter/non-rate-limited-api.png" %}}
+
+When endpoints are not rate-limited and can be called endlessly, it can lead to issues like heavy resource usage, server timeouts, and unfair resource allocation.
+
+To avoid these problems and secure our APIs, it's crucial to implement rate limiting for our endpoints. This ensures that users cannot make an excessive number of requests within a specific time frame. To achieve this, we'll create a rate-limiter middleware for our endpoints.
+
+### Step 3: Configure the Rate limit
+To configure a rate limiter for our application endpoints, paste the following code into the `rateLimiter.js` file:
+```js
+const setRateLimit = require("express-rate-limit");
+
+// Rate limit middleware
+const rateLimitMiddleware = setRateLimit({
```

Comment:
Does `express-rate-limit` also offer other rate-limiting algorithms?

Reply:
I actually had to do more research on this. It looks like I was wrong at first the `express-rate-limit` did not specify the exact rate-limiting algorithm it uses in its documentation or official repository. 

I have made modifications to my previous statement. Thank you